### PR TITLE
[WIP] fix wd, replace pickle with misc pickle

### PR DESCRIPTION
**Pull request by antinucleon**
@piiswrong

### CODE REVIEW

**Comment by piiswrong** [comment]
It's probably better to make it work like lr_scale? instead of feeding in names, just feed in wd_scale and assign the scales in model?

**Comment by antinucleon** [comment]
@mavenlin @mli I added pickle which is able to pack classes but not test yet in distributed environment. Cloud you take a look whether it works?

**Comment by antinucleon** [comment]
@piiswrong It is a little bit mass currently, the final goal is to make optimizer aware each parameter, and use a special group of wd/lr/momentum/gamma/schedule for each one. However, the parameter is somehow related to symbol, and our optimizer is not symbol, which makes it a little headache.

File: python/mxnet/model.py [hunk 1]
```diff
@@ -257,7 +258,7 @@ def _train_multi_device(symbol, ctx, arg_names, param_names, aux_names,
                     do_reset = False
                     break
 
-            if do_reset == True:
+            if do_reset is True:
```
Comment (tqchen) [inline python/mxnet/model.py]: simply use ```if do_reset```

File: python/mxnet/ndarray.py [hunk 2]
```diff
@@ -639,7 +639,7 @@ def binary_ndarray_function(lhs, rhs, out=None):
     def unary_ndarray_function(src, out=None):
         """internal NDArray function"""
         if out:
-            if isinstance(out, NDArray) == False:
+            if isinstance(out, NDArray) is False:
```
Comment (tqchen) [inline python/mxnet/ndarray.py]: ```if not isinstance```

**Comment by piiswrong** [comment]
@antinucleon `self.kwargs["arg_names"] = arg_names` this is going to break other optimizers that doesn't have arg_names as argument. I think it should be moved to base Optimizer class as a set_wd_scale method.

**Comment by antinucleon** [comment]
@piiswrong I tested misc.pickle has some problem when pickle optimizer. Ideally, it will pack code together with object, however when it involves heritage, it doesn't work. I will close this PR and send a new PR for the wd issue.

**Comment by piiswrong** [comment]
@antinucleon remember to move arg_names to the base class as a method. It's currently breaking all optimizers other than SGD

**Pull request closed without merging**